### PR TITLE
Add option to display warnings in rustdoc

### LINKED ISSUE #41574: Have an option to output warnings from doctests
Running cargo test runs doctests by default, and shows warnings for regular code, but not for the doctests themselves. This can be enabled by running cargo test -- --nocapture. I'd like to have warnings also show in our CI tests, even for doctests, but setting -- --nocapture shows all the output which we don't want (and can't control, because they come from the underlying libraries complaining about our tests that are supposed to fail).

I'd think having warnings on for doctests by default would be good, but that wouldn't be backwards compatible. Maybe a --doctest-warnings option could be added in lieu of that?

I believe this was done in 5d145c1c18abb1a6b81238dd1b73178635c12f0a, so discussion there might be worth reading. It seems that there are a bunch of warnings that come up with doctests that should be ignored, but there are also ones that can appear that shouldn't be. The question is how do we address both needs?

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Part of #41574.

r? @alexcrichton 

The output for this file:

```rust
/// ```
/// fn foo(x: u32) {}
///
/// foo(2);
/// let x = 1;
/// panic!();
/// ```
fn foo() {}

/// ```
/// fn foo(x: u32) {}
///
/// foo(2);
/// let x = 1;
/// ```
fn foo2() {}

/// ```
/// fn foo(x: u32) {}
///
/// foo(2);
/// let x = 1;
/// panic!();
/// ```
fn foo3() {}

fn main() {
}
```

is the following:

```
> ./build/x86_64-apple-darwin/stage1/bin/rustdoc -Z unstable-options --display-warnings --test test.rs

running 3 tests
test test.rs - foo (line 1) ... FAILED
test test.rs - foo3 (line 18) ... FAILED
test test.rs - foo2 (line 10) ... ok

successes:

---- test.rs - foo2 (line 10) stdout ----
	warning: unused variable: `x`
 --> <anon>:2:8
  |
2 | fn foo(x: u32) {}
  |        ^
  |
  = note: #[warn(unused_variables)] on by default

warning: unused variable: `x`
 --> <anon>:5:5
  |
5 | let x = 1;
  |     ^
  |
  = note: #[warn(unused_variables)] on by default



successes:
    test.rs - foo2 (line 10)

failures:

---- test.rs - foo (line 1) stdout ----
	warning: unused variable: `x`
 --> <anon>:2:8
  |
2 | fn foo(x: u32) {}
  |        ^
  |
  = note: #[warn(unused_variables)] on by default

warning: unused variable: `x`
 --> <anon>:5:5
  |
5 | let x = 1;
  |     ^
  |
  = note: #[warn(unused_variables)] on by default

thread 'rustc' panicked at 'test executable failed:

thread 'main' panicked at 'explicit panic', <anon>:6
note: Run with `RUST_BACKTRACE=1` for a backtrace.

', src/librustdoc/test.rs:317
note: Run with `RUST_BACKTRACE=1` for a backtrace.

---- test.rs - foo3 (line 18) stdout ----
	warning: unused variable: `x`
 --> <anon>:2:8
  |
2 | fn foo(x: u32) {}
  |        ^
  |
  = note: #[warn(unused_variables)] on by default

warning: unused variable: `x`
 --> <anon>:5:5
  |
5 | let x = 1;
  |     ^
  |
  = note: #[warn(unused_variables)] on by default

thread 'rustc' panicked at 'test executable failed:

thread 'main' panicked at 'explicit panic', <anon>:6
note: Run with `RUST_BACKTRACE=1` for a backtrace.

', src/librustdoc/test.rs:317


failures:
    test.rs - foo (line 1)
    test.rs - foo3 (line 18)

test result: FAILED. 1 passed; 2 failed; 0 ignored; 0 measured
```

### CODE REVIEW

**Comment by frewsxcv** [comment]
Travis found a failure

File: src/librustdoc/markdown.rs [hunk 1]
```diff
@@ -166,6 +166,9 @@ pub fn test(input: &str, cfgs: Vec<String>, libs: SearchPaths, externs: Externs,
     old_find_testable_code(&input_str, &mut collector, DUMMY_SP);
     find_testable_code(&input_str, &mut collector, DUMMY_SP);
     test_args.insert(0, "rustdoctest".to_string());
+    if display_warnings {
+        test_args.insert(1, "--display-stdout".to_string());
```
Comment (alexcrichton) [inline src/librustdoc/markdown.rs]: How come this was added? I thought rustc output went to stderr?

File: src/librustdoc/lib.rs [hunk 2]
```diff
@@ -172,6 +172,7 @@ pub fn opts() -> Vec<RustcOptGroup> {
                          or `#![doc(html_playground_url=...)]`",
                         "URL")),
         unstable(optflag("", "enable-commonmark", "to enable commonmark doc rendering/testing")),
+        unstable(optflag("", "display-warnings", "to print code warnings when testing doc")),
```
Comment (alexcrichton) [inline src/librustdoc/lib.rs]: Can you open a tracking issue for this as well?

File: src/librustdoc/markdown.rs (hunk 1 above)
Comment (GuillaumeGomez) [inline src/librustdoc/markdown.rs]: The name of the tag is up to debate. I'm using it to make the difference between warnings and errors, but it's clearly misleading. What name should we give it?

File: src/librustdoc/lib.rs (hunk 2 above)
Comment (GuillaumeGomez) [inline src/librustdoc/lib.rs]: You mean besides the one already opened? (in the PR description)

File: src/librustdoc/markdown.rs (hunk 1 above)
Comment (alexcrichton) [inline src/librustdoc/markdown.rs]: I think I may not even understand what this is doing. What is it doing?

File: src/librustdoc/lib.rs (hunk 2 above)
Comment (alexcrichton) [inline src/librustdoc/lib.rs]: Yes that can suffice, but then this PR should not close that issue.

File: src/librustdoc/markdown.rs (hunk 1 above)
Comment (GuillaumeGomez) [inline src/librustdoc/markdown.rs]: So for now, when running `rustdoc --test`, warnings aren't displayed. Some people want them to be displayed anyway.

Comment (frewsxcv) [inline src/librustdoc/markdown.rs]: Then why not call it `--display-warnings`

Comment (alexcrichton) [inline src/librustdoc/markdown.rs]: @GuillaumeGomez that's not too helpful of a description, though? You're basically just restating the title and description of this PR.

This flag is implemented **inside of libtest** which means that it has nothing to do with doctest warnings, and is an instantly stable interface to all test binaries that we generate, hence my hesitation. If we can avoid it that would be great, but I do not understand what this flag is doing inside of the test harness.

Can you explain what's going on inside of libtest to why this is being passed?

Comment (GuillaumeGomez) [inline src/librustdoc/markdown.rs]: Aaaaah. My bad. So basically, in libtest, if I have the `display-stdout` option enabled, for every non-failing test, I display what the test sent to the output.

Comment (alexcrichton) [inline src/librustdoc/markdown.rs]: Can this be done without exposing a new flag from libtest? Just by modifying internal data structures or using a custom api from libtest?

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by alexcrichton** [comment]
Looks like a good strategy to me, but CI seems to be failing?

**Comment by GuillaumeGomez** [comment]
Yep, forgot to update some pieces of code.

**Pull request merged**